Libraries/LibCore/EventLoop.cpp

```diff
@@ -409,6 +412,91 @@ void EventLoop::post_event(Object& receiver, NonnullOwnPtr<Event>&& event)
     m_queued_events.empend(receiver, move(event));
 }
 
+template <typename RemoveHandler>
+class ProtectSignalHandler
+{
+    AK_MAKE_NONCOPYABLE(ProtectSignalHandler);
+    AK_MAKE_NONMOVABLE(ProtectSignalHandler);
+
+    int& m_handling_signal;
+    int m_handling_signal_prev;
+    RemoveHandler m_remove_handler;
+
+public:
+    ProtectSignalHandler(int& handling_signal, int signal, RemoveHandler remove_handler):
+        m_handling_signal(handling_signal),
+        m_handling_signal_prev(handling_signal),
+        m_remove_handler(move(remove_handler))
+    {
+        handling_signal = signal;
+    }
+
+    ~ProtectSignalHandler()
+    {
+        bool was_removed = (m_handling_signal == 0);
+        m_handling_signal = m_handling_signal_prev;
+        if (was_removed) {
+            // This handler was flagged to be removed
+            m_remove_handler();
+        }
+    }
+};
+
+void EventLoop::handle_signal(int signo)
+{
+    // We need to protect the handler from being removed while handling it
+    ProtectSignalHandler protect(m_handling_signal, signo,
+        [&] {
+           ASSERT(m_handling_signal == signo); // should have been restored
+           unregister_signal(signo);
+        });
+    auto handler = m_signal_handlers.find(signo);
+    if (handler != m_signal_handlers.end()) {
```

Comment:
This does not do what you think it does.
`HashMap::find` looks for hashes, not for keys.

Reply:
Nevermind, that's the two-argument form.
the one-argument form looks for keys, my bad.

Reply:
> Did you mean to revert [dec27e5](https://github.com/SerenityOS/serenity/commit/dec27e5e6fad25fedd9203087b8107242369dd2b)?
> I don't think [5cdb0ef](https://github.com/SerenityOS/serenity/commit/5cdb0ef2e5bd741e82073069fe8a5ae1b29ec0ff) messes with anything interesting, it just skips setting pgrp to get the tty-related signals while the child is running
Hm well it seemed that this new `m_is_accepting_signals` flag was trying to work mask the issue, but maybe that commit is really two separate things?

Reply:
Not quite, `m_is_accepting_signals` is trying to mask away signals that the shell doesn't care about (e.g. SIGINT and SIGHUP) while a child process is running.
That's really to fix this interaction:
```sh
> cd
# change terminal size
^D
> # Shell now has the wrong idea about the terminal size
```